common/src/main/java/io/pravega/common/concurrent/ScheduledQueue.java

```diff
@@ -0,0 +1,347 @@
+/**
+ * Copyright Pravega Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.pravega.common.concurrent;
+
+import com.google.common.collect.Iterators;
+import java.util.AbstractQueue;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map.Entry;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.ConcurrentSkipListMap;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+import javax.annotation.Nonnull;
+import lombok.Data;
+import lombok.val;
+
+/**
+ * Provides a unbounded blocking queue which {@link Scheduled} items can be added to.
+ * Items which are scheduled will not be returned from {@link #poll()} or {@link #take()} until their scheduled time.
+ */
+public class ScheduledQueue<E extends Scheduled> extends AbstractQueue<E> implements BlockingQueue<E> {
```

Comment:
Just curious, what's the difference between ScheduledQueue and [DelayQueue](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/DelayQueue.html), they look pretty similar to me. Thanks!

Reply:
https://github.com/pravega/pravega/issues/6145#issuecomment-879453547

Reply:
I also added a comment.